Dynamics products and prices to Shopify

charlotta-frends opened the pull request:
#96 

Checklist for review:
- [x] Process name is fine, uses normal casing and descriptive
- [x] Process description is fine, uses normal casing, dots and is descriptive
  - [x] Element names use normal casing, no unneeded capital letters.
- [x] Correct casing used for formats and terms, e.g. CSV, FTP, JSON, Avro
  - [x] Element names do not contain dots
  - [x] Comments / note element text are in good language and end with dots.
- [x] Template parameters checked
  - [x] Parameter descriptions are good and end with dots.
  - [x] Sensitive parameters are marked as secrets
- [x] "Turn off logging" switch is on for any task that has sensitive data as output
- [x] End "balls" like error throws and return statements have titles (double-click) and reasonable information.
  - [x] Results have good summaries, preferably with statistics of some sort like "processed X files"
  - [x] Errors have some error info, usually from task (e.g. some ErrorMessage property contains additional error info)
- [x] Process is organised with groups (dashed boxes)
- [x] Straight lines are preferred where possible
- [x] Subprocesses are forbidden
- [x] Text mode is preferred over Expressions in expressions, e.g. instead of *$"This is {#var.error}"* use *This is {{#var.error}}*
- [x] Process metadata file is present and content is fine (version number and name)
- [x] Process export file is present (process.json)
- [x] Process description file is present (all sections filled in)
- [x] Process description has a SVG screenshot of the process
- [x] Tempalte files are going to the correct folder


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a template for syncing products and prices from Dynamics 365 to Shopify, including error notifications for failed uploads.
  - Added metadata specifying details for the Dynamics 365 to Shopify integration.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

FrendsSarlinS reviewed (changes requested):
1) Is it possible to include an error message to the first Throw-statement? Dynamics-task seems to return an error message, if it fails.

2) The return-statement that ends the whole process returns a single number, if no errors are present. Could we include an additional message?

3) Not necessarily an issue, mainly something to think about, but something is bugging me with the "Find Shopify product id" code element. Thinking about keeping the template as low-code as possible, could the element be replaced with a foreach loop already existing in Frends?
3.1) Again just me being curious, but is it even possible for the included if-statement to return null, i.e. title exists but SKU doesn't?

4) It might be possible to tidy out the process by removing the expressions "Increase updated count" and "Increase inserted count". Replacing them with returns including the increment would do the same trick.

All in all, the process looks neat and functions correctly as far as my tests go! Also, you should consider if Coderabbit's suggestions are worth committing to the long-description.

charlotta-frends commented:
> 1. Is it possible to include an error message to the first Throw-statement? Dynamics-task seems to return an error message, if it fails.

Added the error message from the task to the first throw.

> 2. The return-statement that ends the whole process returns a single number, if no errors are present. Could we include an additional message?

Changed the end return message of the whole process to be a bit more verbal.

> 3. Not necessarily an issue, mainly something to think about, but something is bugging me with the "Find Shopify product id" code element. Thinking about keeping the template as low-code as possible, could the element be replaced with a foreach loop already existing in Frends?

Switched the code block to the ready-made foreach element, although it added extra steps to the process (needed to assign the productId variable separately for it to work). Not sure if it's generally better to favor as much low-code as possible, or as little steps as possible, but at least here I thought the low-code approach seemed cleaner, and also I think it may run a little bit faster.

>    3.1) Again just me being curious, but is it even possible for the included if-statement to return null, i.e. title exists but SKU doesn't?

The "return null;" was only there because every path in that code block needed to return something. Maybe not the cleanest solution, as the foreach basically always returns the sku. :D

> 4. It might be possible to tidy out the process by removing the expressions "Increase updated count" and "Increase inserted count". Replacing them with returns including the increment would do the same trick.

Switched the code expressions to returns according to the suggestion.

FrendsSarlinS commented:
I noticed one more thing, not sure if it was in the previous version.

1) For the new foreach, I'd suggest replacing the "Find Shopify product id" code block with an exclusive decision, since it only contains an if-statement. The yes-route return would #var.productId = #var.shopifyProduct.id, whereas the no-route could maybe result in a return-error?

Other than that, it looks great to me! I ticked the rest of the checkboxes about errors and return statements.

charlotta-frends commented:
> 1. For the new foreach, I'd suggest replacing the "Find Shopify product id" code block with an exclusive decision, since it only contains an if-statement. The yes-route return would #var.productId = #var.shopifyProduct.id, whereas the no-route could maybe result in a return-error?

@FrendsSarlinS The problem with an exclusive decision here is that if a matching SKU is not found with the first product, the loop should not do/return anything, and it only should move on to check the next one, and so on. With the exclusive decision it would return the error and end the loop without checking the rest of the products.

I initially didn't put any error check after the foreach, because the check after the update task should catch it anyway (with productId as null). But would it be better to add one? And/or I could clarify the function of the foreach loop in the text annotation a bit more.

FrendsSarlinS commented:
Ah, I see! It's odd it would do that if you use return-statement in the loop (or the loop inside a loop), as it should check the next product with a matching title. That's why we use returns in the first place, so the whole process wouldn't shut down in case of an error! 😆 But seeing that there is an issue with that, I can see why you opted to use the code block.

Regarding the error catching, I'd say it depends on whether it's relevant or more pleasant for the user that the process gives a detailed error message as of what and when (i.e. "update failed because productId is null" or something like that) or not. If the PUT request (the update task) fails, the task would send an exception instead of a return message, so I'd say the error check would be better option here.

charlotta-frends commented:
@FrendsSarlinS My bad, the suggestion to switch the code expression to exclusive decision worked just fine, so I made the change! I also added the error check after the inner foreach for any change the matching SKU is not found.

charlotta-frends commented:
Another commit in, added some missing return element names.

FrendsSarlinS reviewed (approved):
With these changes, it's good to go by me!

charlotta-frends commented:
Updated the Dynamics task after the task bugfix. Also made some updates according to the review comments from my other templates.

FrendsSarlinS commented on Templates/ERPs/Dynamics365/Dynamics products and prices to Shopify/metadata.json:
```diff
@@ -0,0 +1,6 @@
+  "name": "Dynamics products and prices to Shopify",
+  "version": "1.0.0",
+  "tags": [],
+  "additionalSystemsUsed": []
```
Mention Shopify here.

charlotta-frends replied:
Done
